src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlClientMetricDiagnosticListener.cs

```diff
@@ -0,0 +1,88 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using System.Reflection;
+
+namespace OpenTelemetry.Instrumentation.SqlClient.Implementation;
+
+#if NET6_0_OR_GREATER
+internal class SqlClientMetricDiagnosticListener : ListenerHandler
+{
+    internal static readonly Assembly Assembly = typeof(SqlClientMetricDiagnosticListener).Assembly;
+    internal static readonly AssemblyName AssemblyName = Assembly.GetName();
+    internal static readonly string MeterName = AssemblyName.Name!;
+    internal static readonly string MeterVersion = AssemblyName.Version!.ToString();
+    internal static readonly Meter Meter = new(MeterName, MeterVersion);
+    private static readonly Histogram<double> DbClientOperationDuration = Meter.CreateHistogram<double>(
+        "db.client.operation.duration",
+        "s",
+        "Duration of database client operations.");
+
+#if !NET7_0_OR_GREATER
+    private static readonly double StopwatchTickFrequency = (double)TimeSpan.TicksPerSecond / Stopwatch.Frequency;
+#endif
+
+    private readonly AsyncLocal<long> beginTimestamp = new AsyncLocal<long>();
+
+    public SqlClientMetricDiagnosticListener(string sourceName)
+        : base(sourceName)
+    {
+    }
+
+    public override void OnEventWritten(string name, object? payload)
+    {
+        var activity = Activity.Current;
+        switch (name)
+        {
+            case SqlClientDiagnosticListener.SqlDataBeforeExecuteCommand:
+            case SqlClientDiagnosticListener.SqlMicrosoftBeforeExecuteCommand:
+                {
+                    // only record the start time if tracing is disabled or sampling has occured
+
+                    if (activity == null)
+                    {
+                        this.beginTimestamp.Value = Stopwatch.GetTimestamp();
+                    }
+
+                    break;
+                }
+
+            case SqlClientDiagnosticListener.SqlDataAfterExecuteCommand:
+            case SqlClientDiagnosticListener.SqlMicrosoftAfterExecuteCommand:
+                {
+                    TagList tags = default;
+                    tags.Add(new("db.system", "mssql"));
+
+                    if (activity == null)
+                    {
+                        if (this.beginTimestamp.Value != 0)
+                        {
+                            var endTimestamp = Stopwatch.GetTimestamp();
+                            var duration = GetElapsedTime(this.beginTimestamp.Value, endTimestamp);
+                            DbClientOperationDuration.Record(duration.TotalSeconds, tags);
+                        }
+                    }
+                    else
+                    {
+                        activity.Stop();
```

Comment:
Related to the other comment: I think stopping this activity here is a bit brutal: this activity is not really started by this class - it's actually "owned" by `SqlClientDiagnosticListener` and it expects this to also stop. In the current form, it'll trigger this log: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlClientDiagnosticListener.cs#L153

Now.. I guess it's stopped here, because otherwise there is no `activity.Duration`, right? I think this again could be solved by handling the events in a common class and not subscribing twice to the same event.